### PR TITLE
[Improvement] we can pick how lax we are during AvroSink output valid…

### PR DESCRIPTION
**Pull request by mproch**
We can control how strict avro sink is when it comes to validation:
- should we check that optional fields are present (even if they're null)
- should we check there are not redundant parameters?

TODO: handle default values properly

### CODE REVIEW

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import pl.touk.nussknacker.engine.api.typed.{CanBeSubclassDeterminer, SingleSubclassRestriction, typing}
+import pl.touk.nussknacker.engine.api.typed.typing.{TypedObjectTypingResult, TypingResult}
+import pl.touk.nussknacker.engine.avro.KafkaAvroBaseTransformer.SinkValueParamName
+import pl.touk.nussknacker.engine.avro.typed.AvroSchemaTypeDefinitionExtractor
+
+object OutputValidator {
+
+  def validateOutput(value: TypingResult, schema: Schema, validationMode: ValidationMode)(implicit nodeId: NodeId): Validated[CustomNodeError, Unit] = {
+    val possibleTypes = AvroSchemaTypeDefinitionExtractor.ExtendedPossibleTypes
+    //FIXME: this still does not handle optional fields validation properly for acceptUnfilledOptional == true?
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala:16]: 1. Can you change it to less offensive `TODO`, because it doens't look like it is very "hot" problem
2. I think that it is something missing in this comment. Am I right that you mean "not handle defined optional fields type type validation"? Or something like this? Because now it looks like we not handle this validation at all...

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    val returnType = new AvroSchemaTypeDefinitionExtractor(skippNullableFields = validationMode.acceptUnfilledOptional).typeDefinition(schema, possibleTypes)
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala:17]: I know that it is not part of your change, but `skipNullableFields` is wrong name because we skip `optional` fields. By optional we should understand field with default value. It can be filed without union with null but with default value.

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+  val strict: ValidationMode = ValidationMode("strict", "Strict mode", acceptUnfilledOptional = false, acceptRedundant = false)
+
+  val allowOptional: ValidationMode = ValidationMode("allowOptional", "No optional parameters validations", acceptUnfilledOptional = true, acceptRedundant = false)
+
+  val allowRedundantAndOptional: ValidationMode =  ValidationMode("allowRedundantAndOptional", "Allow redundant parameters", acceptUnfilledOptional = true, acceptRedundant = true)
+
+  val values: List[ValidationMode] = List(strict, allowOptional, allowRedundantAndOptional)
+
+  def byName(name: String): Option[ValidationMode] = values.find(_.name == name)
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala:15]: I don't think that byName will be a good idea to identify mode, because we can change this label. WDYT about having in parameters value `(acceptUnfilledOptional,acceptRedundant)` or just  `acceptUnfilledOptional,acceptRedundant`?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactory.scala [hunk 4]
```diff
@@ -50,19 +53,26 @@ class KafkaAvroSinkFactory(val schemaRegistryProvider: SchemaRegistryProvider, v
   private def initialParametersForNode(implicit nodeId: NodeId): WriterT[Id, List[ProcessCompilationError], List[Parameter]] =
     topicParam.map(List(
       Parameter.optional[CharSequence](KafkaAvroBaseTransformer.SinkKeyParamName).copy(isLazyParameter = true),
-      Parameter[AnyRef](KafkaAvroBaseTransformer.SinkValueParamName).copy(isLazyParameter = true), _))
+      Parameter[AnyRef](KafkaAvroBaseTransformer.SinkValueParamName).copy(isLazyParameter = true),
+      //TODO: default value
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactory.scala:57]: First mode won't be the default one?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    val restriction = new SingleSubclassRestriction {
+      override def canBeSubclassOf(givenClass: typing.SingleTypingResult, superclassCandidate: typing.SingleTypingResult): Boolean = (givenClass, superclassCandidate) match {
+        case (TypedObjectTypingResult(objFields, _), TypedObjectTypingResult(superFields, _)) if !validationMode.acceptRedundant =>
+          objFields.keys.forall(superFields.keys.toSet.contains)
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala:21]: Nice idea, I like this approach - no new mode introduced :+1:

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactory.scala [hunk 6]
```diff
@@ -50,19 +53,26 @@ class KafkaAvroSinkFactory(val schemaRegistryProvider: SchemaRegistryProvider, v
   private def initialParametersForNode(implicit nodeId: NodeId): WriterT[Id, List[ProcessCompilationError], List[Parameter]] =
     topicParam.map(List(
       Parameter.optional[CharSequence](KafkaAvroBaseTransformer.SinkKeyParamName).copy(isLazyParameter = true),
-      Parameter[AnyRef](KafkaAvroBaseTransformer.SinkValueParamName).copy(isLazyParameter = true), _))
+      Parameter[AnyRef](KafkaAvroBaseTransformer.SinkValueParamName).copy(isLazyParameter = true),
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactory.scala:56]: I wonder if better approach would be having this param in "endingParametersForNode". Because now you will got this parameter at the beginning and you force user to make some hard decision. If it would be at the end with some reasonable default, in 99% user won't be worried of fact that this parameter exist.

Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactory.scala:56]: Well, ok, but OTOH it seems good to have it close to output...

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactory.scala (hunk 4 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactory.scala:57]: yes, true

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala (hunk 2 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala:17]: ok

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala (hunk 1 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala:16]: ok

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala (hunk 3 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala:15]: currently I have name and label to avoid such problems :)

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/BestEffortAvroEncoder.scala [hunk 7]
```diff
@@ -136,8 +136,8 @@ class BestEffortAvroEncoder(avroSchemaEvolution: AvroSchemaEvolution) {
     fields.asScala.map {
       case (fieldName, value) =>
         val field = schema.getField(fieldName)
-        if (field == null) {
-          error(s"Not expected field with name: $fieldName for schema: $schema")
+        if (field == null && !validationMode.acceptRedundant) {
+          error(s"Not expected field with name: $fieldName for schema: $schema and policy $validationMode does not allow redundant")
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/BestEffortAvroEncoder.scala:140]: How it will work? Can you add some test for that? In case of `field == null && validationMode.acceptRedundant` I see NPE in here: `field.schema()` :)

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package pl.touk.nussknacker.engine.avro.encode
+
+case class ValidationMode private(name: String, label: String, acceptUnfilledOptional: Boolean, acceptRedundant: Boolean)
+
+object ValidationMode {
+
+  val strict: ValidationMode = ValidationMode("strict", "Strict mode", acceptUnfilledOptional = false, acceptRedundant = false)
```
Comment (lciolecki) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala:7]: these all static field should be also final ?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala (hunk 5 above)
Comment (arkadius) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala:22]: After thinking a while:
1. Shouldn't bidirectional check: `value.canBeSubclassOf(returnType) && returnType.canBeSubclassOf(value)` will be enough?
2. Maybe better approch would be just extends some method `checkObjectFields` in `CanBeSubclassDeterminer` instead of creating new abstraction. Current solution will be complicated in the future when we will want to change `&& restriction....` to `|| restriction` or sth similar

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala (hunk 3 above)
Comment (arkadius) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala:15]: Indeed I see now. IMO `acceptUnfilledOptional,acceptRedundant` will be better because it will be easier to migrate such value in case when our validation approach will slightly changed e.g. one flag will be replaced by another or we will add other combinations of modes ... but it is also fine :+1:

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala (hunk 8 above)
Comment (arkadius) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/ValidationMode.scala:7]: @lciolecki In scala, final modifier for object's vals do nothing. `final` mean "won't be overridden" `val` mean "is contant"

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala (hunk 5 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala:22]: 1. It also came to my mind, but I think cases like: String vs CharSequence, Int vs Long etc. will cause problems 
2. good idea, I'll try it.

Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/OutputValidator.scala]: done

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/BestEffortAvroEncoder.scala (hunk 7 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/encode/BestEffortAvroEncoder.scala]: done...
